cql3/query_processor.cc

```diff
@@ -1056,9 +1056,17 @@ query_processor::execute_schema_statement(const statements::schema_altering_stat
             co_await stmt.grant_permissions_to_creator(client_state);
         }
     }
-
-    // If an IF [NOT] EXISTS clause was used, this may not result in an actual schema change.  To avoid doing
-    // extra work in the drivers to handle schema changes, we return an empty message in this case. (CASSANDRA-7600)
+    if (!ce) {
+        // If an IF [NOT] EXISTS clause was used and resource was already created
+        // we don't emit schema change event. However it interacts badly with
+        // concurrent clients creating resources. The client seeing no create event
+        // assumes resource already previously existed and proceeds with its logic
+        // which may depend on that resource. But it may send requests to nodes which
+        // are not yet aware of new schema or client's metadata may be outdated.
+        // To force synchronization we'll try to emit fake event (see
+        // github.com/scylladb/scylladb/issues/16909).
+        ce = stmt.created_event();
```

Comment:
I don't understand why we call created_event both here and in the statement itself. Wouldn't it be simpler to create the event unconditionally in the statement, or here, and not conditionally both here and in the statement?

Reply:
yes, to make it easier created_event function was extracted, so one need to list all usages of this function.

Reply:
Wouldn't it be simpler to call created_event unconditionally, and only here?

btw what happens if we don't create and IF NOT EXISTS was not specified?

Reply:
> btw what happens if we don't create and IF NOT EXISTS was not specified? 

statement code throws exception

> Wouldn't it be simpler to call created_event unconditionally, and only here?

yes, maybe but then:
- it's hard to revert once we fix the underlying problem
- other events are still spread around

Reply:
What would a fix to the underlying problem look like?

Reply:
I don't think it's practical; it relies on cooperation with drivers.

The server-side approach of always sending the event has no downsides (well, CREATE statements are slower) and work with all drivers.

Reply:
it doesn't really fix more generic case - schema change are not immediate from driver point of view, if you have a workflow like that:
- driver instance A alters schema
- driver instance A sends to driver instance B internal message (i.e. user's RPC)
- driver instance B assumes finished alter statement effects are present (on every node)
- driver instance B receives a failure as the assumption doesn't hold for all nodes

Reply:
@avikivity can we resolve this comment and queue the patch or you have further comments?

Reply:
> it doesn't really fix more generic case - schema change are not immediate from driver point of view, if you have a workflow like that:
> 
> * driver instance A alters schema

At this point it waits for schema agreement

> * driver instance A sends to driver instance B internal message (i.e. user's RPC)
> * driver instance B assumes finished alter statement effects are present (on every node)
> * driver instance B receives a failure as the assumption doesn't hold for all nodes

Why does it not hold?

Also, how is this related to CREATE TABLE IF NOT EXISTS?

Reply:
> At this point it waits for schema agreement

Unless it's  statement with `if not exists`. But maybe there are other cases in some drivers?

Previously it wasn't easy to do what you ask due to auto-grant, but I improved it in https://github.com/scylladb/scylladb/pull/18819/commits/d5697cf5ccfdea289df04b9f9deb733c127af83e

So now it's possible to return create event unconditionally in a single place. It feels wrong though, and we need to be careful to not create conditional logic on the event (similar to https://github.com/scylladb/scylladb/blob/master/cql3/query_processor.cc#L1038). But it's doable so if you insist I can refactor it.